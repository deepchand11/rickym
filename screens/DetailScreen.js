import React, { useEffect, useState } from "react";
import {
  StyleSheet,
  Text,
  View,
  ActivityIndicator,
  Button,
} from "react-native";

const DetailScreen = ({ route, navigation }) => {
  const [detail, setDetail] = useState({});
  const [loading, setLoading] = useState(false);
  const { url } = route.params;

  useEffect(() => {
    (async () => {
      try {
        setLoading(true);
        const response = await fetch(url);
        const data = await response.json();
        if (data) {
          setLoading(false);
          setDetail(data);
        }
      } catch (err) {
        console.log(err);
      }
    })();
    return () => {
      setDetail({});
    };
  }, []);

  return (
    <View style={styles.container}>
      {loading ? (
        <ActivityIndicator color="black" style={{ margin: 15 }} />
      ) : (
        <View>
          <Text style={styles.label}>Name:</Text>
          <Text style={styles.text} testID="detailNameId">
            {detail.name}
          </Text>
          <Text style={styles.label}>Type:</Text>
          <Text style={styles.text} testID="detailTypeId">
            {detail.type}
          </Text>
          <Text style={styles.label}>Dimension:</Text>
          <Text style={styles.text} testID="detailDimensionId">
            {detail.dimension}
          </Text>
        </View>
      )}

      <Button
        testID="backBtn"
        title="Go back"
        onPress={() => navigation.popToTop()}
      />
    </View>
  );
};

export default DetailScreen;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#E4E4E4",
    padding: 20,
  },
  label: {
    color: "rgb(158, 158, 158)",
    fontWeight: "bold",
    fontSize: 14,
  },
  text: {
    fontSize: 16,
    marginBottom: 10,
  },
});
