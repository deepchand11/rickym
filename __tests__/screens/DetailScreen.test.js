import "react-native";
import React from "react";
import DetailScreen from "../../screens/DetailScreen";
import { fireEvent, waitFor, render, act } from "@testing-library/react-native";

// import { create, act } from "react-test-renderer";
// const popToTop = jest.fn();
// const urlMock = {
//   params: {
//     url: "locationUrl",
//   },
// };
// const tree = create(<DetailScreen route={urlMock} navigation={{ popToTop }} />);
// test("snapshot", () => {
//   expect(tree).toMatchSnapshot();
// });

// test("call timeout", () => {
//   act(() => jest.runAllTimers());
// });
// Note: test renderer must be required after react-native.
// import renderer from "react-test-renderer";

it("Detail Screen renders correctly", async () => {
  const popToTop = jest.fn();
  const urlMock = {
    params: {
      url: "locationUrl",
    },
  };
  const { toJSON } = await render(
    <DetailScreen route={urlMock} navigation={{ popToTop }} />
  );

  expect(toJSON()).toMatchSnapshot();
});

it("Navigate to previous screen", async () => {
  const popToTop = jest.fn();
  const urlMock = {
    params: {
      url: "locationUrl",
    },
  };
  const { getByTestId } = await render(
    <DetailScreen route={urlMock} navigation={{ popToTop }} />
  );
  const backBtn = getByTestId("backBtn");
  fireEvent.press(backBtn);
  expect(popToTop).toHaveBeenCalledTimes(1);
});

// it("Api mock calling on component mount ", async () => {
//   const popToTop = jest.fn();
//   const urlMock = {
//     params: {
//       url: "locationUrl",
//     },
//   };
//   const success = {
//     data: { name: "Mockname", type: "Mocktype", dimension: "Mockdimension" },
//   };
//   const fetchEpisode = fetch.mockResponseSuccess(JSON.stringify(success));
//   await act(async () => {
//     const { getByTestId } = await render(
//       <DetailScreen route={urlMock} navigation={{ popToTop }} />
//     );
//     expect(getByTestId("episodeNameId").props.children).toBe(success.data.name);
//   });
// });

it("Api mock calling on component mount success", async () => {
  const popToTop = jest.fn();
  const urlMock = {
    params: {
      url: "locationUrl",
    },
  };
  const success = {
    name: "Mockname",
    type: "Mocktype",
    dimension: "Mockdimension",
  };
  const fetchLocation = fetch.mockResponseSuccess(JSON.stringify(success));
  await act(async () => {
    const { getByTestId } = await render(
      <DetailScreen route={urlMock} navigation={{ popToTop }} />
    );
    expect(getByTestId("detailNameId").props.children).toBe("Mockname");
    expect(getByTestId("detailTypeId").props.children).toBe("Mocktype");
    expect(getByTestId("detailDimensionId").props.children).toBe(
      "Mockdimension"
    );
  });
  await waitFor(() => expect(fetchLocation).toHaveBeenCalledTimes(1));
});

it("Api mock calling on component mount error", async () => {
  const popToTop = jest.fn();
  const urlMock = {
    params: {
      url: "locationUrl",
    },
  };
  const error = {
    message: "Mockerror",
  };
  const fetchLocation = fetch.mockResponseFailure(error);
  await act(async () => {
    const { getByTestId } = await render(
      <DetailScreen route={urlMock} navigation={{ popToTop }} />
    );
    expect(getByTestId("detailNameId").props.children).toBeUndefined();
    expect(getByTestId("detailTypeId").props.children).toBeUndefined();
    expect(getByTestId("detailDimensionId").props.children).toBeUndefined();
  });
  await waitFor(() => expect(fetchLocation).toHaveBeenCalledTimes(1));
});
