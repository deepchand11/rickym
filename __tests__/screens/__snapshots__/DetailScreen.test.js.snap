// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail Screen renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#E4E4E4",
      "flex": 1,
      "padding": 20,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="black"
    hidesWhenStopped={true}
    size="small"
    style={
      Object {
        "margin": 15,
      }
    }
  />
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="backBtn"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Go back
      </Text>
    </View>
  </View>
</View>
`;
